test/controllers/csvs_controller_test.rb

```diff
@@ -0,0 +1,335 @@
+        end
+      end
+    end
+
+    describe "a GET to show for invalid" do
+      setup { get :show, id: -1, format: format }
+
+      describe "as html" do
+        let(:format) { :html }
+
+        it "redirects to index with error flash" do
+          assert_redirected_to csvs_path
+          flash[:error].must_equal "The CSV export does not exist."
+        end
+      end
+
+      describe "as JSON" do
+        let(:format) { :json }
+
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response 404
+          @response.body.must_equal "{\"status\":\"not-found\"}"
+        end
+      end
+    end
+
+    describe "a GET to show for finished" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
+        @csv_export.finished!
+        get :show, id: @csv_export.id, format: format
+      end
+
+      describe "as html" do
+        let(:format) { :html }
+
+        it "renders the status template" do
+          assert_template :show
+          @response.body.must_include "download"
+          @response.body.must_include csv_path(id: @csv_export.id, format: 'csv')
+        end
+      end
+
+      describe "as JSON" do
+        let(:format) { :json }
+
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response :ok
+          @csv_export.reload
+          @response.body.must_equal @csv_export.to_json
+        end
+      end
+    end
+
+    describe "a GET to show for downloaded" do
+      setup do
+        @csv_export = CsvExport.find(export.id)
+        @csv_export.downloaded!
+        get :show, id: @csv_export.id, format: format
+      end
+
+      describe "as html" do
+        let(:format) { :html }
+
+        it "renders the status template" do
+          assert_template :show
+          @response.body.must_include "download"
+          @response.body.must_include csv_path(id: @csv_export.id, format: 'csv')
+        end
+      end
+
+      describe "as JSON" do
+        let(:format) { :json }
+
+        it "renders json" do
+          assert_equal "application/json", @response.content_type
+          assert_response :ok
+          @csv_export.reload
+          @response.body.must_equal @csv_export.to_json
+        end
+      end
+    end
+  end
+
+  as_a_deployer do
+    describe "a POST to create" do
+      teardown do
+        @csv_exports = CsvExport.where(user_id: deployer.id)
+        @csv_exports.each do |csv_file|
+          filename = "#{Rails.root}/export/#{csv_file.id}"
+          File.delete(filename) if File.exist?(filename)
+          csv_file.delete
+        end
+      end
+    
+      it "should create a new csv_export and redirects to status" do
+        assert_difference 'CsvExport.count' do
+          post :create, content: "deploys"
+        end
+        assert_redirected_to csv_path(CsvExport.last)
+      end
+    end
+
+    describe "a Get to show as csv with file" do
```

Comment:
might be simpler to consolidate all the show tests into 1 block ... then it's easier to see what is going on and usually works is being done on 1 method so only that methods block needs to be inspected and not all of the test file